dll/win32/shlwapi/utils.cpp

```diff
@@ -35,65 +35,174 @@
+
+        CHAR dir[MAX_PATH];
+        if (pwszDir)
+        {
+            info.fMask |= CMIC_MASK_UNICODE;
+            info.lpDirectoryW = pwszDir;
+            if (SHUnicodeToAnsi(pwszDir, dir, _countof(dir)))
+                info.lpDirectory = dir;
+        }
+
+        hr = pCM->InvokeCommand((LPCMINVOKECOMMANDINFO)&info);
     }
 
-    /* Invoking itself doesn't need the menu object, but getting the command info
-       needs the menu. */
+    if (pUnk)
+        IUnknown_SetSite(pCM, NULL);
     if (hMenu)
         DestroyMenu(hMenu);
 
-    SetCursor(hOldCursor);
+    return hr;
+}
 
-    return ret;
+/*************************************************************************
+ * SHInvokeCommandOnContextMenuEx [SHLWAPI.639]
+ */
+EXTERN_C
+HRESULT WINAPI
+SHInvokeCommandOnContextMenuEx(
+    _In_opt_ HWND hWnd,
+    _In_opt_ IUnknown* pUnk,
+    _In_ IContextMenu* pCM,
+    _In_ UINT fCMIC,
+    _In_ UINT fCMF,
+    _In_opt_ LPCSTR pszVerb,
+    _In_opt_ LPCWSTR pwszDir)
+{
+    return SHInvokeCommandOnContextMenuInternal(hWnd, pUnk, pCM, fCMIC, fCMF, pszVerb, pwszDir, true);
+}
+
+/*************************************************************************
+ * SHInvokeCommandOnContextMenu [SHLWAPI.540]
+ */
+EXTERN_C
+HRESULT WINAPI
+SHInvokeCommandOnContextMenu(
+    _In_opt_ HWND hWnd,
+    _In_opt_ IUnknown* pUnk,
+    _In_ IContextMenu* pCM,
+    _In_ UINT fCMIC,
+    _In_opt_ LPCSTR pszVerb)
+{
+    return SHInvokeCommandOnContextMenuEx(hWnd, pUnk, pCM, fCMIC, CMF_EXTENDEDVERBS, pszVerb, NULL);
+}
+
+/*************************************************************************
+ * SHInvokeCommandWithFlagsAndSite [SHLWAPI.571]
+ */
+EXTERN_C
+HRESULT WINAPI
+SHInvokeCommandWithFlagsAndSite(
+    _In_opt_ HWND hWnd,
+    _In_opt_ IUnknown* pUnk,
+    _In_ IShellFolder* pShellFolder,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ UINT fCMIC,
+    _In_opt_ LPCSTR pszVerb)
+{
+    HRESULT hr = E_INVALIDARG;
+    if (pShellFolder)
+    {
+        IContextMenu *pCM;
+        hr = pShellFolder->GetUIObjectOf(hWnd, 1, &pidl, IID_IContextMenu, NULL, (void**)&pCM);
+        if (SUCCEEDED(hr))
+        {
+            fCMIC |= CMIC_MASK_FLAG_LOG_USAGE;
+            hr = SHInvokeCommandOnContextMenuEx(hWnd, pUnk, pCM, fCMIC, 0, pszVerb, NULL);
+            pCM->Release();
+        }
+    }
+    return hr;
+}
+
+
+/*************************************************************************
+ * IContextMenu_Invoke [SHLWAPI.207]
+ *
+ * Used by Win:SHELL32!CISFBand::_TrySimpleInvoke.
+ */
+EXTERN_C
+BOOL WINAPI
+IContextMenu_Invoke(
+    _In_ IContextMenu *pContextMenu,
+    _In_ HWND hwnd,
+    _In_ LPCSTR lpVerb,
+    _In_ UINT uFlags)
+{
+    TRACE("(%p, %p, %s, %u)\n", pContextMenu, hwnd, debugstr_a(lpVerb), uFlags);
+    HRESULT hr = SHInvokeCommandOnContextMenuInternal(hwnd, NULL, pContextMenu, 0,
+                                                      uFlags, lpVerb, NULL, false);
+    return SUCCEEDED(hr);
```

Comment:
Same here, I'd like to have error logging if something goes wrong.
I prefer one error being logged in multiple places over it being not logged at all.

If there is some issue, and all failing functions are completely silent it is making debugging a lot harder.